match2/commons/match_group_display_bracket.lua

```diff
@@ -1,738 +1,745 @@
+        for _, x in ipairs(match.bracketData.lowerMatches) do
+            upperMatchIds[x.matchId] = matchId
         end
     end
 
-    _bracketConfig = _getBracketConfig(args)
+    -- Computes the layout of a match and everything to its left.
+    local computeNodeLayout = FnUtil.memoizeY(function(matchId, computeNodeLayout)
+        local match = matchesById[matchId]
+        local lowerLayouts = Array.map(
+            match.bracketData.lowerMatches,
+            function(x) return computeNodeLayout(x.matchId) end
+        )
 
-    local matches, referencedIds = _mapMatches(matches, has3rd)
-    _matches = matches
+        -- Compute partial sums of heights of lower round matches
+        local heightSums = LuaUtils.math.partialSums(
+            Array.map(lowerLayouts, function(layout) return layout.height end)
+        )
 
-    out = html.create('div'):addClass('brkts-main')
-    firsthead = true
-    height = 0
-    for id, match in utils.iter.spairs(matches) do
-        if not referencedIds[match.id] then
-            local extradata, node = unpack(match:buildBracket(firsthead))
-            out:node(node)
-            height = height + math.max(extradata[1], extradata[1] / 2 + match.minDisplayHeight)
-            firsthead = false
+        -- Show a connector line without joints if there is a single lower round 
+        -- match advancing an opponent that is placed near the middle of this match.
+        local singleStraightLine
+        if #lowerLayouts == 1 then
+            local opponentIx = match.bracketData.lowerMatches[1].opponentIx
+            singleStraightLine = 
+                #match.opponents % 2 == 0 
+                    and (opponentIx == #match.opponents / 2 or opponentIx == #match.opponents / 2 + 1)
+                or #match.opponents % 2 == 1
+                    and opponentIx == math.floor(#match.opponents / 2) + 1
+        else
+            singleStraightLine = false
         end
-    end
-    
-    out:css('min-height', height .. 'px')
-
-    return out
-end
 
-function _getBracketConfig(args)
-    local globalConfig = Json.parse(tostring(mw.message.new('BracketConfig')))
-    local config = {}
-    for param, format in pairs(BRACKET_CONFIG_FORMAT) do
-        local val = utils.misc.emptyOr(args[param], globalConfig[param], BRACKET_CONFIG_DEFAULT[param])
-        if val ~= nil then
-            if format == 'number' then
-                config[param] = tonumber(val)
-            elseif format == 'boolean' then
-                config[param] = utils.misc.readBool(val)
-            elseif format == 'string' then
-                config[param] = tostring(val)
+        -- Don't show the header if it's disabled. Also don't show the header 
+        -- if it is the first match of a round and if a higher round match can
+        -- show it instead.
+        local isFirstChild = upperMatchIds[matchId] and matchId == matchesById[upperMatchIds[matchId]].bracketData.lowerMatches[1].matchId
```

Comment:
The variable name is incorrect, right? It's not just the first child, it also checks if a higher round match can show it. So would be good to split this up. Once that's done, you should also be able to remove the comment

Reply:
I reworded the comment slightly. Basically as long as headers aren't disabled via `config.hideRoundTitles`, the first lower match advancing into a higher match can have its headers shown in the higher match. So the checks are just whether it's the first child, whether it has a header, and whether headers are disabled.